ASoC: SOF: core/Intel: Follow the pause_supported flag from topology to disable or keep the pause support enabled

ujfalusi opened the pull request:
Hi,

new token set is added by https://github.com/thesofproject/sof/pull/9202 to indicate that the pause operation is supported or not on the given PCM device.  

Pause is an optional feature that depends on pipeline, topology and modules running within that pipeline. Lets have a marker to help userspace and kernel know which pipelines should advertise pause.

Use this information in hda-pcm.c to disable the pause support or keep it enabled.

ujfalusi committed:
ASoC: SOF: Add support for pause supported tokens from topology

New tokens are added to topology:
1202: SOF_TKN_STREAM_PLAYBACK_PAUSE_SUPPORTED
1203: SOF_TKN_STREAM_CAPTURE_PAUSE_SUPPORTED

The new tokens are used to advertise support for PAUSE/RESUME operation on
a PCM device depending on firmware product, use case, pipeline topology.

The snd_sof_pcm_stream.pause_supported is updated to reflect the advertised
value for the PCM device.

If the token does not exist then the pause_supported is set to false.

Note: it is up to the platform code to use this flag to decide to advertise
the PAUSE support for user space or not.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: hda-pcm: Follow the pause_supported flag to drop PAUSE support

If the stream's pause_supported flag is false then mask out the PAUSE
support, so user space will be prevented to use it.

Introduce a module parameter to ignore the pause_supported flag, named as
force_pause_support to allow testing of the PAUSE feature.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Yes, with this PR the PAUSE is going to be mostly disabled for Intel platforms regardless of IPC version.

kv2019i reviewed (commented):
Typoe in title "enabeld"

Patch looks good but I wonder if this is ok to change the behaviour for existing systems -- this will go all the way back to CML/CFL systems (and not just Intel).. It's possible somebody is happily using SOF on a system with apps that use pause, and this kernel PR could break their use-case when they upgrade the kernel.

Staged introduction (or disable_pause token in tplg) would be more incremental.

plbossart commented:
I am not aligned with the approach, sorry.

I agree that PAUSE is not a release blocker. but in practice it's the only way to trigger quick transitions and find bugs we wouldn't see otherwise.

It's also incorrect to say it's not been tested. What has not been tested is suspend while paused which is a second-order corner case.

The net effect of this topology token would be to COMPLETELY disable PAUSE, including in our CI tests. That's different to disabling it for releases, and making an educated decision to avoid spending cycles on problematic issues with no ROI.

The topology approach has zero flexibiliity, it's all or nothing.

I wouldn't mind if this topology-based approach was contingent on a 'default n' kernel Kconfig, which would only be used by  developers and CI tests. I am not ok with completely disabling this PAUSE capability for tests.

ujfalusi commented:
> Typoe in title "enabeld"
> 
> Patch looks good but I wonder if this is ok to change the behaviour for existing systems -- this will go all the way back to CML/CFL systems (and not just Intel).

Right now the suspend while paused is causing the system to hard freeze. I have no idea when this regression happened, but it shows the extent to testing and use of pause/resume. Sure, this is a corner case, but I have a faint recollection that it worked.
The only solution so far for the freeze is my https://github.com/thesofproject/linux/pull/5040, for both IPC3 and IPC4.

> It's possible somebody is happily using SOF on a system with apps that use pause, and this kernel PR could break their use-case when they upgrade the kernel.

Yes, everything is a possibility, true.
 
> Staged introduction (or disable_pause token in tplg) would be more incremental.

Initially I wanted to have opt-out for pause/resume but https://github.com/thesofproject/linux/issues/5035 shows that this feature has not been tested by CI and users to gave any confidence that it is working.

ujfalusi commented:
Changes since v1:
- add `force_pause_support` module parameter (to hda-pcm.c) to ignore the pasue_supported flag from topology and to keep the pause enabled. Iow, if the `force_pause_support` is set to `true` then there is no change in behavior

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -407,6 +407,10 @@ static const struct sof_topology_token stream_tokens[] = {
+	{SOF_TKN_STREAM_PLAYBACK_PAUSE_SUPPORTED, SND_SOC_TPLG_TUPLE_TYPE_BOOL, get_token_u16,
+		offsetof(struct snd_sof_pcm, stream[0].pause_supported)},
+	{SOF_TKN_STREAM_CAPTURE_PAUSE_SUPPORTED, SND_SOC_TPLG_TUPLE_TYPE_BOOL, get_token_u16,
+		offsetof(struct snd_sof_pcm, stream[1].pause_supported)},
```
Do we need to have two tokens?

What if we just had one SOF_TKN_STREAM_PCM_PAUSE_SUPPORTED?

Presumably we already know the direction of the stream.

But maybe it's required because we don't have a means to apply the same value in the two directions?

ujfalusi replied:
I have used the D0I3 support as template.
this is for the PCM device, it can be duplex or playback only or capture only.
It might be possible that only the playback is allowed to support pause while the capture is not in case of a duplex PCM device.

plbossart replied:
ok

plbossart commented:
@fredoh9 @marc-hb @ssavati can we add this to the CI configuration before merging this:

````
options snd_sof_intel_hda_common force_pause_support=true
````

plbossart reviewed (dismissed):
Sounds good, let's make sure CI is ready before merging this, otherwise we're going to get a slew of errors across the board.

ujfalusi commented:
Changes since v2:
- drop the`hda_` prefix for the module parameter name (no idea how it keeps coming back - removed like four times while playing with it...)

marc-hb commented:
> but https://github.com/thesofproject/linux/issues/5035 shows that this feature has not been tested by CI and users to gave any confidence that it is working.

#5035 is a pretty specific corner case. Simpler use cases have been regularly tested.

marc-hb commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -240,6 +245,14 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
 		runtime->hw.info &= ~SNDRV_PCM_INFO_PAUSE;
 
+	/*
+	 * Follow the hint (unless a module parameter is set to ignore it) and
```
Why is this called a "hint" when it's always followed? Because older kernel versions ignore it?

This could/should be clearer.

ujfalusi replied:
The flag is documented in previous patch as 'hint' as the flag is defaulting to false for all platforms, it is only handled by Intel code and older topologies will have PAUSE disabled also.
This is intentional decision, but I rather call it as a hint in SOF context than a hard rule, I don't expect NXP/AMD/Mediatek to disable PAUSE support, but they can if they also set to follow this hint.

marc-hb replied:
Sorry I should have spent more time reading the commit message. A commit message is not a substitute for documentation though, is there or will there be any? Somewhere in ALSA user space maybe?

A "hint" is indeed something that may or may not have an effect depending on other factors. However this particular "hint" will be 100% deterministic on every device! It will be 100% effective on some devices and 100% ignored on other devices. It is a "hint" only from the perspective of the developer (=you) looking at all devices at once. But it is not a hint from the perspective of any final user. I don't think this should be called a "hint", it's not necessary and quite misleading. It's just a parameter that is not supported in every configuration. Ignoring some (e.g. incompatible) parameter in some configurations is not so unusual and not called a "hint".


> I don't expect NXP/AMD/Mediatek to disable PAUSE support, but they can if they also set to follow this hint.

What would be great is to print a warning in all configurations where the parameter is set and ignored, is that possible? Silently ignoring user input is extremely time consuming.

ujfalusi commented:
Changes since v3:
- Reword commit messages and comments to omit the 'hint' word.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -332,6 +332,14 @@ struct snd_sof_pcm_stream {
 	bool d0i3_compatible; /* DSP can be in D0I3 when this pcm is opened */
+	/*
+	 * Flag for the core and platform code about PCM pause/resume support.
+	 * Note: ENd-users are using audio via media servers (Pulseaudio, Pipewire,
```
stray capital letter: End-users

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -332,6 +332,14 @@ struct snd_sof_pcm_stream {
+	 * Flag for the core and platform code about PCM pause/resume support.
+	 * Note: ENd-users are using audio via media servers (Pulseaudio, Pipewire,
+	 * CRAS, etc) and media servers do not use pause/resume.
+	 * Media players using the ALSA directly have means to handle PCM devices
```
using the ALSA API directly?

ujfalusi replied:
or 'using ALSA directly' ?

plbossart commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -153,6 +153,8 @@
 #define SOF_TKN_STREAM_PLAYBACK_COMPATIBLE_D0I3	1200
 #define SOF_TKN_STREAM_CAPTURE_COMPATIBLE_D0I3	1201
+#define SOF_TKN_STREAM_PLAYBACK_PAUSE_SUPPORTED	1202
+#define SOF_TKN_STREAM_CAPTURE_PAUSE_SUPPORTED	1203
```
commit message:
SOF_TKN_STREAM_PLAYBACK_CAPTURE_SUPPORTED ->  SOF_TKN_STREAM_CAPTURE_PAUSE_SUPPORTED

ujfalusi replied:
gash, thanks for catching it.

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -240,6 +245,14 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
+	 */
+	if (!hda_force_pause_support &&
+	    !spcm->stream[substream->stream].pause_supported)
+		runtime->hw.info &= ~SNDRV_PCM_INFO_PAUSE;
```
it's a bit weird because we could have a configuration with hda_always_enable_dmi_l1=1 and hda_force_pause_support=1, and in this case the force_pause_support does not work..

ujfalusi replied:
in case of forced  dmi_l1 we cannot enable the pause support as it would lead to a hardware issue.
Currently the hda_always_enable_dmi_l1 would "silently" disable the pause support for only capture.

I'm not sure how is better to handle this...

plbossart replied:
maybe add a condition to check that hda_always_enable_dmi_l1 and hda_force_pause_support have compatible values.

ujfalusi replied:
@plbossart, if they are not compatible? I think we should always obey the `hda_always_enable_dmi_l1` and disable the pause support for capture regardless of what `hda_force_pause_support` commands, like I have done it.

plbossart reviewed (commented):
couple of corrections needed @ujfalusi but no objections on the directions.

ujfalusi commented:
Changes since v4:
- Fix the commit message and comment in the first patch

marc-hb commented:
> Right now the suspend while paused is causing the system to hard freeze. I have no idea when this regression happened, but it shows the extent to testing and use of pause/resume. Sure, this is a corner case, but I have a faint recollection that it worked.

This does not "show" or prove anything: it's only a corner case that regular users are incredibly unlikely to hit - much less likely than regular "pause" which is already unused.

We actually have a test script for this corner case but it's never been working properly:
- https://github.com/thesofproject/sof-test/pull/1222

On the other hand, regular pause/release (without suspend) has been regularly tested since forever and it has been mostly passing. There were some frequent false positives caused by bugs in the test code but I fixed all that a few days ago in rewrite https://github.com/thesofproject/sof-test/pull/1218 and pause/release testing has been rock solid since.

> I agree that PAUSE is not a release blocker. but in practice it's the only way to trigger quick transitions and find bugs we wouldn't see otherwise.

Indeed: since my test code rewrite in 1218, only LNL has been failing consistently. Guess what: LNL is the only platform where we ALSO have other, persistent and hard to fix failures - but with much lower reproduction rates. Coincidence? I don't think so. 

- https://github.com/thesofproject/sof/issues/9191
- https://github.com/thesofproject/sof/issues/9164
- https://github.com/thesofproject/sof/issues/9123
- https://github.com/thesofproject/sof/issues/8809
- https://github.com/thesofproject/linux/issues/5109
- https://github.com/thesofproject/linux/issues/5086
- https://github.com/thesofproject/linux/issues/5080
- https://github.com/thesofproject/linux/issues/5109

It's very common (and very useful) to stress test systems past supported use cases to **merely increase reproduction rates of issues also happening in regular use cases** and this has what pause/release has been successfully doing for a long time - and even better now.

marc-hb commented:
BTW what is the `aplay/arecord` user interface expectation after this is merged? I mean for regular users, not for CI or experts with `force_pause_support=true` or something.

https://sof-ci.01.org/linuxpr/PR5041/build4054/devicetest/index.html?model=GLK_BOB_DA7219-ipc3&testcase=check-pause-resume-playback-10 and others show this:

```
t=150 ms: aplay: (0/10) Found volume ### | xx%, recording for 178 ms
t=234 ms: aplay: (0/10) Found   === PAUSE ===  ,  pausing for 195 ms
t=404 ms: aplay: (1/10) Found volume ### | xx%, recording for 161 ms
t=404 ms: aplay: WARNING: received == PAUSE == while in state recording! Ignoring.
t=575 ms: aplay: (1/10) Found   === PAUSE ===  ,  pausing for 128 ms
t=746 ms: aplay: (2/10) Found volume ### | xx%, recording for 151 ms
t=746 ms: aplay: WARNING: received == PAUSE == while in state recording! Ignoring.
t=831 ms: aplay: (2/10) Found   === PAUSE ===  ,  pausing for 186 ms
t=1002 ms: aplay: (3/10) Found volume ### | xx%, recording for 179 ms
t=1002 ms: aplay: WARNING: received == PAUSE == while in state recording! Ignoring.
t=1258 ms: aplay: (3/10) Found   === PAUSE ===  ,  pausing for 124 ms
t=1343 ms: aplay: (4/10) Found volume ### | xx%, recording for 150 ms
t=1343 ms: aplay: WARNING: received == PAUSE == while in state recording! Ignoring.
t=1514 ms: aplay: (4/10) Found   === PAUSE ===  ,  pausing for 175 ms
t=1684 ms: aplay: (5/10) Found volume ### | xx%, recording for 111 ms
t=1684 ms: aplay: WARNING: received == PAUSE == while in state recording! Ignoring.
...
```

This is really weird: why does `aplay/arecord` display PAUSE when there is no PAUSE support?

Maybe ALSA userspace should be involved in this too?

ujfalusi commented:
@marc-hb, something is not quite right in the expect script in case-lib/apause.exp, not sure what to be honest.

ujfalusi commented:
@marc-hb, I think I have found a plausible reason...
aplay prints out the `=== PAUSE ===` unconditionally before trying to pause. If pause is not supported then it will write `PAUSE command ignored (no hw support)\n` which will replace the previously printed `=== PAUSE ===`
expect might be catching this.
https://github.com/alsa-project/alsa-utils/blob/master/aplay/aplay.c#L1645
https://github.com/alsa-project/alsa-utils/blob/master/aplay/aplay.c#L1610

So the expect script is OK.

marc-hb commented:
> In every cases it is using the same line

Yes, aplay uses `\r` instead of `\n` to constantly overwrite its previous output and not scroll. The script should be ready for that. If you think it's not please file an new sof-test bug and I'll try to fix ASAP.

Generally speaking, Expect (or anything reading from a pipe) has no concept of "overwriting" or "scrolling"; only terminals do. Other programs reading from a pipe only see a stream of characters, some of them happening to be `\r` and `\n`.

> aplay prints out the `=== PAUSE ===` unconditionally before trying to pause. If pause is not supported then it will write `PAUSE command ignored (no hw support)` which will replace the previously printed === PAUSE ===
expect might be catching this.

Thanks for the investigation!

> So the expect script is OK.

The script is unfortunately not OK because it trusts the first, unconditional (and TBH: _wrong_) `=== PAUSE ===` message and changes its state machine accordingly. And then it gets very confused.

Not sure how to fix that. Is there some sort of more convenient API to detect the lack of pause support? And then abort/skip the test immediately instead of wasting a lot of time running pointlessly and returning misleading results. I mean more convenient than this very misleading aplay output.

ujfalusi commented:
> The script is unfortunately not OK because it trusts the first, unconditional (and TBH: _wrong_) `=== PAUSE ===` message and changes its state machine accordingly. And then it gets very confused.
> 
> Not sure how to fix that. Is there some sort of more convenient API to detect the lack of pause support? And then abort/skip the test immediately instead of wasting a lot of time running pointlessly and returning misleading results. I mean more convenient than this very misleading aplay output.

I'm not sure about other ways to check runtime info flag of the PCM, but that would be ideal

This might avoid confusing the script (might need to handle the not supported case and abort?): https://github.com/alsa-project/alsa-utils/pull/271

marc-hb commented:
> I'm not sure about other ways to check runtime info flag of the PCM, but that would be ideal

Anyone?

> This might avoid confusing the script (might need to handle the not supported case and abort?): https://github.com/alsa-project/alsa-utils/pull/271

Very useful but it won't help systems and people who don't update ALSA user space regularly.

> The script is unfortunately not OK because it trusts the first, unconditional (and TBH: wrong) === PAUSE === message and changes its state machine accordingly. And then it gets very confused. Not sure how to fix that. 

Wait: it shouldn't be too hard to catch `PAUSE command ignored` and fail or skip even after entering a "confused" state. It only needs to happen before the confusion leads to another, earlier and confusing failure...

plbossart commented:
@ujfalusi @marc-hb I am not able to follow what the thread is about. Is our CI able to deal with this or not? 

The agreement was that the pause support would be forced on all CI devices, so what is the issue? We don't really care about the case where the pause is NOT supported, because it's not really our problem is it?

What I don't get is whether all CI devices are ready for this PR to be merged, adding a kernel parameter requires some changes in the setup.

marc-hb commented:
> We don't really care about the case where the pause is NOT supported, because it's not really our problem is it?

It's not our problem in theory. In practice, we still don't have consistent device configuration.

marc-hb commented:
> Is our CI able to deal with this or not?

It will be after small sof-test fix https://github.com/thesofproject/sof-test/pull/1226 is merged. That fix will let us tell the difference between misconfigured devices versus well configured devices actually failing the test. @ujfalusi please review.

marc-hb commented:
SOFCI TEST

lgirdwood reviewed (changes requested):
@ujfalusi can you rewrite the commit message. Thanks.

ujfalusi commented:
Changes since v3:
- Commit message updated
- The module parameter updated from bool to int with three value: -1: Use default, 0L Disable pause, 1: Enable pause

ssavati commented:
@ujfalusi I have added below flag in all CI devices
`options snd_sof_intel_hda_common force_pause_support=true`

ujfalusi commented:
SOFCI TEST

lgirdwood commented:
SOFCI TEST
